From 7c9f606d11fe8d77747aa1d8f10de35b69665df9 Mon Sep 17 00:00:00 2001
From: liyong4x <yongx.a.li@intel.com>
Date: Tue, 12 Jun 2018 21:13:29 -0400
Subject: [PATCH 096/129] media: intel-ipu4: ar0231at:HDR mode compatible with
 linear mode

Message for Open Source:
Ar0231at supports Linear exposure mode, 2-HDR mode, 3-HDR mode and
4-HDR mode. Driver should support all functionalities sensor has.
So add all these four exposure mode driver codes.

Message for Internal:
Linear exposure mode and 2-HDR, 3-HDR, 4-HDR mode would be compatible.
User must select the exposure mode before starting stream.

[Issue/Feature]
Driver should provide the controlling methods to configure
the exposure mode besides linear mode and all HDR modes.

[Root Cause/Changes]
1. Add the linear mode and 2-HDR, 3-HDR, 4-HDR initial settings.
2. Selet ctrl_data_modes when control v4l2 with CRL_CID_EXPOSURE_MODE.
3. Change the impact mode as CRL_IMPACTS_MODE_SELECTION.
4. Change the default fll and llp value as the default 3-HDR mode.

Change-Id: I1ebbbf0fa78923d96e95a442b7824e22ce17891d
Tracked-On: #JIIAP-475
Signed-off-by: liyong4x <yongx.a.li@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 .../i2c/crlmodule/crl_ar0231at_configuration.h     | 1124 +++++++++++++++++++-
 1 file changed, 1104 insertions(+), 20 deletions(-)

diff --git a/drivers/media/i2c/crlmodule/crl_ar0231at_configuration.h b/drivers/media/i2c/crlmodule/crl_ar0231at_configuration.h
index 96626fc..3b35854 100644
--- a/drivers/media/i2c/crlmodule/crl_ar0231at_configuration.h
+++ b/drivers/media/i2c/crlmodule/crl_ar0231at_configuration.h
@@ -68,7 +68,33 @@ struct crl_subdev_rect_rep ar0231at_1920_1080_rects[] = {
 	}
 };
 
-static struct crl_register_write_rep ar0231at_1920_1080[] = {
+/*
+ * Exposure mode:
+ *   0: Linear mode
+ *   1: 2-HDR mode
+ *   2: 3-HDR mode
+ *   3: 4-HDR mode
+ */
+struct crl_ctrl_data_pair ar0231at_ctrl_data_modes[] = {
+	{
+		.ctrl_id = CRL_CID_EXPOSURE_MODE,
+		.data = 0,
+	},
+	{
+		.ctrl_id = CRL_CID_EXPOSURE_MODE,
+		.data = 1,
+	},
+	{
+		.ctrl_id = CRL_CID_EXPOSURE_MODE,
+		.data = 2,
+	},
+	{
+		.ctrl_id = CRL_CID_EXPOSURE_MODE,
+		.data = 3,
+	},
+};
+
+static struct crl_register_write_rep ar0231at_1920_1080_linear_mode[] = {
 	{ 0x301A, CRL_REG_LEN_16BIT, 0x1058, 0x10 },
 	{ 0x0000, CRL_REG_LEN_DELAY, 200, 0x10 },
 	{ 0x3092, CRL_REG_LEN_16BIT, 0x0C24, 0x10 },
@@ -360,23 +386,1036 @@ struct crl_subdev_rect_rep ar0231at_1920_1080_rects[] = {
 	{ 0x3400, CRL_REG_LEN_16BIT, 0x0010, 0x10 },
 	{ 0x3402, CRL_REG_LEN_16BIT, 0x0F10, 0x10 },
 	{ 0x3404, CRL_REG_LEN_16BIT, 0x0970, 0x10 },
-	{ 0x3082, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
-	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F1, 0x10 },
-	{ 0x0000, CRL_REG_LEN_DELAY, 200, 0x10 },
-	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F0, 0x10 },
-	{ 0x300C, CRL_REG_LEN_16BIT, 0x0872, 0x10 },
-	{ 0x300A, CRL_REG_LEN_16BIT, 0x054A, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F1, 0x10 },
+	{ 0x0000, CRL_REG_LEN_DELAY, 200, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F0, 0x10 },
+	{ 0x300C, CRL_REG_LEN_16BIT, 0x0872, 0x10 },
+	{ 0x300A, CRL_REG_LEN_16BIT, 0x054A, 0x10 },
+	{ 0x3042, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3012, CRL_REG_LEN_16BIT, 0x0163, 0x10 },
+	{ 0x3014, CRL_REG_LEN_16BIT, 0x014F, 0x10 },
+	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C08, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x31D0, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x31AE, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x31AC, CRL_REG_LEN_16BIT, 0x0C0C, 0x10 },
+	/* try sync mode */
+	{ 0x340A, CRL_REG_LEN_16BIT, 0x0077, 0x10 },
+	{ 0x340C, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
+	{ 0x30CE, CRL_REG_LEN_16BIT, 0x0120, 0x10 },
+	{ 0x301A, CRL_REG_LEN_16BIT, 0x19DC, 0x10 },
+};
+
+static struct crl_register_write_rep ar0231at_1920_1080_2hdr_mode[] = {
+	{ 0x301A, CRL_REG_LEN_16BIT, 0x10D8, 0x10 },
+	{ 0x0000, CRL_REG_LEN_DELAY, 100, 0x10 },
+	{ 0x3092, CRL_REG_LEN_16BIT, 0x0C24, 0x10 },
+	{ 0x337A, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x3520, CRL_REG_LEN_16BIT, 0x1288, 0x10 },
+	{ 0x3522, CRL_REG_LEN_16BIT, 0x880C, 0x10 },
+	{ 0x3524, CRL_REG_LEN_16BIT, 0x0C12, 0x10 },
+	{ 0x352C, CRL_REG_LEN_16BIT, 0x1212, 0x10 },
+	{ 0x354A, CRL_REG_LEN_16BIT, 0x007F, 0x10 },
+	{ 0x350C, CRL_REG_LEN_16BIT, 0x055C, 0x10 },
+	{ 0x3506, CRL_REG_LEN_16BIT, 0x3333, 0x10 },
+	{ 0x3508, CRL_REG_LEN_16BIT, 0x3333, 0x10 },
+	{ 0x3100, CRL_REG_LEN_16BIT, 0x4000, 0x10 },
+	{ 0x3280, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3282, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3284, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3286, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3288, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x328A, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x328C, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x328E, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3290, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3292, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3294, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3296, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3298, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x329A, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x329C, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x329E, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x301A, CRL_REG_LEN_16BIT, 0x10D8, 0x10 },
+	{ 0x0000, CRL_REG_LEN_DELAY, 200, 0x10 },
+	{ 0x2512, CRL_REG_LEN_16BIT, 0x8000, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0905, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3350, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2004, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1460, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1578, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x7B24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xEA24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1022, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2410, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x155A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1400, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x24FF, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x24FF, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x24EA, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2324, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x647A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2404, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x052C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x400A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF0A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF0A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1008, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3851, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1440, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0004, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0801, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0408, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1180, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2652, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1518, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0906, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1348, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1002, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1016, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1181, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1189, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1056, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0D09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1413, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2B15, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0311, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD909, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1214, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4109, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0312, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1409, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0110, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD612, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1212, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xDD11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD910, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5609, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1511, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xDB09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1511, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x9B09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0F11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBB12, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1A12, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1014, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5010, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x7610, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xE609, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0812, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x290B, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0904, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1440, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0923, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x15C8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x13C8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x092C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1588, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1388, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0C09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0C14, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4109, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1112, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6212, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBF11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBB10, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6611, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFB09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3511, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBB12, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6312, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6014, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0015, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xB812, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xA012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0010, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2610, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0013, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3053, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4215, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4013, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4010, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1611, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8111, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8910, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5612, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x010D, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0815, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC015, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD013, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1313, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0215, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC015, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC813, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0515, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8813, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0213, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0411, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC909, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0814, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0109, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD908, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1400, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x091A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1440, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0903, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1214, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x10D6, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1212, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11DD, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11D9, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1056, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0917, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11DB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0913, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11FB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0905, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x121A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1460, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1250, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1076, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x10E6, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x15A8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x13A8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1240, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1260, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0925, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x13AD, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0902, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0907, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1588, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x138D, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0914, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B13, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1C0C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0920, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1262, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1260, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BF, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1066, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x090A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11FB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x093B, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1263, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1260, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1400, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1508, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11B8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x12A0, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1200, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1026, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1000, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1300, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1100, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x437A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0609, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B05, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0708, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4137, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x502C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2CFE, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x15FE, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0C2C, 0x10 },
+	{ 0x32E6, CRL_REG_LEN_16BIT, 0x00E0, 0x10 },
+	{ 0x1008, CRL_REG_LEN_16BIT, 0x036F, 0x10 },
+	{ 0x100C, CRL_REG_LEN_16BIT, 0x058F, 0x10 },
+	{ 0x100E, CRL_REG_LEN_16BIT, 0x07AF, 0x10 },
+	{ 0x1010, CRL_REG_LEN_16BIT, 0x014F, 0x10 },
+	{ 0x3230, CRL_REG_LEN_16BIT, 0x0312, 0x10 },
+	{ 0x3232, CRL_REG_LEN_16BIT, 0x0532, 0x10 },
+	{ 0x3234, CRL_REG_LEN_16BIT, 0x0752, 0x10 },
+	{ 0x3236, CRL_REG_LEN_16BIT, 0x00F2, 0x10 },
+	{ 0x3566, CRL_REG_LEN_16BIT, 0x3328, 0x10 },
+	{ 0x32D0, CRL_REG_LEN_16BIT, 0x3A02, 0x10 },
+	{ 0x32D2, CRL_REG_LEN_16BIT, 0x3508, 0x10 },
+	{ 0x32D4, CRL_REG_LEN_16BIT, 0x3702, 0x10 },
+	{ 0x32D6, CRL_REG_LEN_16BIT, 0x3C04, 0x10 },
+	{ 0x32DC, CRL_REG_LEN_16BIT, 0x370A, 0x10 },
+	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
+	{ 0x302A, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x302C, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x302E, CRL_REG_LEN_16BIT, 0x0003, 0x10 },
+	{ 0x3030, CRL_REG_LEN_16BIT, 0x004E, 0x10 },
+	{ 0x3036, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x3038, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
+	{ 0x30A2, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x30A6, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x3040, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3040, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0004, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F1, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x3180, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
+	{ 0x33E4, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x3004, CRL_REG_LEN_16BIT, 0x0004, 0x10 },
+	{ 0x3008, CRL_REG_LEN_16BIT, 0x0783, 0x10 },
+	{ 0x3002, CRL_REG_LEN_16BIT, 0x0040, 0x10 },
+	{ 0x3006, CRL_REG_LEN_16BIT, 0x0477, 0x10 },
+	{ 0x3032, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3400, CRL_REG_LEN_16BIT, 0x0010, 0x10 },
+	{ 0x3402, CRL_REG_LEN_16BIT, 0x0788, 0x10 },
+	{ 0x3402, CRL_REG_LEN_16BIT, 0x0F10, 0x10 },
+	{ 0x3404, CRL_REG_LEN_16BIT, 0x04B8, 0x10 },
+	{ 0x3404, CRL_REG_LEN_16BIT, 0x0970, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0004, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F1, 0x10 },
+	{ 0x300C, CRL_REG_LEN_16BIT, 0x07BA, 0x10 },
+	{ 0x300A, CRL_REG_LEN_16BIT, 0x05C8, 0x10 },
+	{ 0x3042, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3012, CRL_REG_LEN_16BIT, 0x0163, 0x10 },
+	{ 0x3014, CRL_REG_LEN_16BIT, 0x0882, 0x10 },
+	{ 0x321E, CRL_REG_LEN_16BIT, 0x0882, 0x10 },
+	{ 0x3222, CRL_REG_LEN_16BIT, 0x0882, 0x10 },
+	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C0E, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C0E, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C0E, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x31D0, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x31AE, CRL_REG_LEN_16BIT, 0x0201, 0x10 },
+	{ 0x31AE, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x31AC, CRL_REG_LEN_16BIT, 0x140C, 0x10 },
+	{ 0x340A, CRL_REG_LEN_16BIT, 0x0077, 0x10 },
+	{ 0x340C, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
+	{ 0x30CE, CRL_REG_LEN_16BIT, 0x0120, 0x10 },
+	{ 0x301A, CRL_REG_LEN_16BIT, 0x19DC, 0x10 },
+};
+
+static struct crl_register_write_rep ar0231at_1920_1080_3hdr_mode[] = {
+	{ 0x301A, CRL_REG_LEN_16BIT, 0x10D8, 0x10 },
+	{ 0x0000, CRL_REG_LEN_DELAY, 100, 0x10 },
+	{ 0x3092, CRL_REG_LEN_16BIT, 0x0C24, 0x10 },
+	{ 0x337A, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x3520, CRL_REG_LEN_16BIT, 0x1288, 0x10 },
+	{ 0x3522, CRL_REG_LEN_16BIT, 0x880C, 0x10 },
+	{ 0x3524, CRL_REG_LEN_16BIT, 0x0C12, 0x10 },
+	{ 0x352C, CRL_REG_LEN_16BIT, 0x1212, 0x10 },
+	{ 0x354A, CRL_REG_LEN_16BIT, 0x007F, 0x10 },
+	{ 0x350C, CRL_REG_LEN_16BIT, 0x055C, 0x10 },
+	{ 0x3506, CRL_REG_LEN_16BIT, 0x3333, 0x10 },
+	{ 0x3508, CRL_REG_LEN_16BIT, 0x3333, 0x10 },
+	{ 0x3100, CRL_REG_LEN_16BIT, 0x4000, 0x10 },
+	{ 0x3280, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3282, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3284, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3286, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3288, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x328A, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x328C, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x328E, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3290, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3292, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3294, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3296, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3298, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x329A, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x329C, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x329E, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x301A, CRL_REG_LEN_16BIT, 0x10D8, 0x10 },
+	{ 0x0000, CRL_REG_LEN_DELAY, 200, 0x10 },
+	{ 0x2512, CRL_REG_LEN_16BIT, 0x8000, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0905, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3350, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2004, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1460, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1578, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x7B24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xEA24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1022, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2410, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x155A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1400, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x24FF, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x24FF, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x24EA, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2324, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x647A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2404, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x052C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x400A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF0A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF0A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1008, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3851, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1440, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0004, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0801, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0408, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1180, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2652, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1518, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0906, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1348, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1002, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1016, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1181, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1189, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1056, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0D09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1413, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2B15, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0311, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD909, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1214, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4109, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0312, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1409, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0110, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD612, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1212, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xDD11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD910, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5609, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1511, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xDB09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1511, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x9B09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0F11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBB12, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1A12, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1014, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5010, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x7610, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xE609, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0812, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x290B, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0904, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1440, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0923, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x15C8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x13C8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x092C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1588, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1388, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0C09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0C14, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4109, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1112, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6212, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBF11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBB10, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6611, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFB09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3511, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBB12, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6312, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6014, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0015, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xB812, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xA012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0010, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2610, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0013, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3053, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4215, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4013, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4010, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1611, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8111, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8910, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5612, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x010D, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0815, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC015, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD013, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1313, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0215, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC015, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC813, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0515, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8813, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0213, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0411, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC909, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0814, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0109, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD908, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1400, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x091A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1440, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0903, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1214, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x10D6, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1212, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11DD, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11D9, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1056, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0917, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11DB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0913, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11FB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0905, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x121A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1460, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1250, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1076, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x10E6, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x15A8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x13A8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1240, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1260, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0925, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x13AD, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0902, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0907, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1588, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x138D, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0914, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B13, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1C0C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0920, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1262, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1260, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BF, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1066, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x090A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11FB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x093B, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1263, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1260, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1400, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1508, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11B8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x12A0, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1200, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1026, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1000, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1300, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1100, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x437A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0609, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B05, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0708, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4137, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x502C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2CFE, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x15FE, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0C2C, 0x10 },
+	{ 0x32E6, CRL_REG_LEN_16BIT, 0x00E0, 0x10 },
+	{ 0x1008, CRL_REG_LEN_16BIT, 0x036F, 0x10 },
+	{ 0x100C, CRL_REG_LEN_16BIT, 0x058F, 0x10 },
+	{ 0x100E, CRL_REG_LEN_16BIT, 0x07AF, 0x10 },
+	{ 0x1010, CRL_REG_LEN_16BIT, 0x014F, 0x10 },
+	{ 0x3230, CRL_REG_LEN_16BIT, 0x0312, 0x10 },
+	{ 0x3232, CRL_REG_LEN_16BIT, 0x0532, 0x10 },
+	{ 0x3234, CRL_REG_LEN_16BIT, 0x0752, 0x10 },
+	{ 0x3236, CRL_REG_LEN_16BIT, 0x00F2, 0x10 },
+	{ 0x3566, CRL_REG_LEN_16BIT, 0x3328, 0x10 },
+	{ 0x32D0, CRL_REG_LEN_16BIT, 0x3A02, 0x10 },
+	{ 0x32D2, CRL_REG_LEN_16BIT, 0x3508, 0x10 },
+	{ 0x32D4, CRL_REG_LEN_16BIT, 0x3702, 0x10 },
+	{ 0x32D6, CRL_REG_LEN_16BIT, 0x3C04, 0x10 },
+	{ 0x32DC, CRL_REG_LEN_16BIT, 0x370A, 0x10 },
+	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
+	{ 0x302A, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x302C, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x302E, CRL_REG_LEN_16BIT, 0x0003, 0x10 },
+	{ 0x3030, CRL_REG_LEN_16BIT, 0x004E, 0x10 },
+	{ 0x3036, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x3038, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
+	{ 0x30A2, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x30A6, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x3040, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3040, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F2, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F2, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F2, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x3180, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
+	{ 0x33E4, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x3004, CRL_REG_LEN_16BIT, 0x0004, 0x10 },
+	{ 0x3008, CRL_REG_LEN_16BIT, 0x0783, 0x10 },
+	{ 0x3002, CRL_REG_LEN_16BIT, 0x0040, 0x10 },
+	{ 0x3006, CRL_REG_LEN_16BIT, 0x0477, 0x10 },
+	{ 0x3032, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3400, CRL_REG_LEN_16BIT, 0x0010, 0x10 },
+	{ 0x3402, CRL_REG_LEN_16BIT, 0x0788, 0x10 },
+	{ 0x3402, CRL_REG_LEN_16BIT, 0x0F10, 0x10 },
+	{ 0x3404, CRL_REG_LEN_16BIT, 0x04B8, 0x10 },
+	{ 0x3404, CRL_REG_LEN_16BIT, 0x0970, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F2, 0x10 },
+	{ 0x300C, CRL_REG_LEN_16BIT, 0x07BA, 0x10 },
+	{ 0x300A, CRL_REG_LEN_16BIT, 0x05C8, 0x10 },
+	{ 0x3042, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3012, CRL_REG_LEN_16BIT, 0x0163, 0x10 },
+	{ 0x3014, CRL_REG_LEN_16BIT, 0x0882, 0x10 },
+	{ 0x321E, CRL_REG_LEN_16BIT, 0x0882, 0x10 },
+	{ 0x3222, CRL_REG_LEN_16BIT, 0x0882, 0x10 },
+	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C0E, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C0E, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C0E, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x31D0, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x31AE, CRL_REG_LEN_16BIT, 0x0201, 0x10 },
+	{ 0x31AE, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x31AC, CRL_REG_LEN_16BIT, 0x140C, 0x10 },
+	{ 0x340A, CRL_REG_LEN_16BIT, 0x0077, 0x10 },
+	{ 0x340C, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
+	{ 0x30CE, CRL_REG_LEN_16BIT, 0x0120, 0x10 },
+	{ 0x301A, CRL_REG_LEN_16BIT, 0x19DC, 0x10 },
+};
+
+static struct crl_register_write_rep ar0231at_1920_1080_4hdr_mode[] = {
+	{ 0x301A, CRL_REG_LEN_16BIT, 0x10D8, 0x10 },
+	{ 0x0000, CRL_REG_LEN_DELAY, 100, 0x10 },
+	{ 0x3092, CRL_REG_LEN_16BIT, 0x0C24, 0x10 },
+	{ 0x337A, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x3520, CRL_REG_LEN_16BIT, 0x1288, 0x10 },
+	{ 0x3522, CRL_REG_LEN_16BIT, 0x880C, 0x10 },
+	{ 0x3524, CRL_REG_LEN_16BIT, 0x0C12, 0x10 },
+	{ 0x352C, CRL_REG_LEN_16BIT, 0x1212, 0x10 },
+	{ 0x354A, CRL_REG_LEN_16BIT, 0x007F, 0x10 },
+	{ 0x350C, CRL_REG_LEN_16BIT, 0x055C, 0x10 },
+	{ 0x3506, CRL_REG_LEN_16BIT, 0x3333, 0x10 },
+	{ 0x3508, CRL_REG_LEN_16BIT, 0x3333, 0x10 },
+	{ 0x3100, CRL_REG_LEN_16BIT, 0x4000, 0x10 },
+	{ 0x3280, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3282, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3284, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3286, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3288, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x328A, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x328C, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x328E, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3290, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3292, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3294, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3296, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x3298, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x329A, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x329C, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x329E, CRL_REG_LEN_16BIT, 0x0FA0, 0x10 },
+	{ 0x301A, CRL_REG_LEN_16BIT, 0x10D8, 0x10 },
+	{ 0x0000, CRL_REG_LEN_DELAY, 200, 0x10 },
+	{ 0x2512, CRL_REG_LEN_16BIT, 0x8000, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0905, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3350, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2004, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1460, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1578, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x7B24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xEA24, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1022, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2410, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x155A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1400, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x24FF, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x24FF, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x24EA, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2324, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x647A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2404, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x052C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x400A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF0A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFF0A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1008, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3851, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1440, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0004, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0801, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0408, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1180, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2652, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1518, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0906, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1348, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1002, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1016, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1181, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1189, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1056, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0D09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1413, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2B15, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0311, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD909, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1214, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4109, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0312, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1409, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0110, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD612, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1212, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xDD11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD910, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5609, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1511, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xDB09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1511, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x9B09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0F11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBB12, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1A12, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1014, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5010, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x7610, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xE609, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0812, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x290B, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0904, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1440, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0923, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x15C8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x13C8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x092C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1588, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1388, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0C09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0C14, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4109, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1112, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6212, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBF11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBB10, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6611, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xFB09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3511, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xBB12, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6312, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x6014, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0015, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xB812, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xA012, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0010, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2610, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0013, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0011, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x3053, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4215, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4013, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4010, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1611, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8111, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8910, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5612, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x010D, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0815, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC015, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD013, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x5009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1313, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0215, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC015, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC813, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0515, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8813, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0213, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0411, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xC909, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0814, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0109, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B11, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0xD908, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1400, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x091A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1440, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0903, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1214, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x10D6, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1212, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11DD, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11D9, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1056, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0917, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11DB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0913, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11FB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0905, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x121A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1210, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1460, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1250, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1076, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x10E6, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x15A8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x13A8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1240, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1260, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0925, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x13AD, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0902, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0907, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1588, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0901, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x138D, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B09, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0914, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4009, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B13, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x8809, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1C0C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0920, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1262, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1260, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BF, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1066, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x090A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11FB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x093B, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11BB, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1263, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1260, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1400, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1508, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x11B8, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x12A0, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1200, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1026, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1000, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1300, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x1100, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x437A, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0609, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0B05, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0708, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x4137, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x502C, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x2CFE, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x15FE, 0x10 },
+	{ 0x2510, CRL_REG_LEN_16BIT, 0x0C2C, 0x10 },
+	{ 0x32E6, CRL_REG_LEN_16BIT, 0x00E0, 0x10 },
+	{ 0x1008, CRL_REG_LEN_16BIT, 0x036F, 0x10 },
+	{ 0x100C, CRL_REG_LEN_16BIT, 0x058F, 0x10 },
+	{ 0x100E, CRL_REG_LEN_16BIT, 0x07AF, 0x10 },
+	{ 0x1010, CRL_REG_LEN_16BIT, 0x014F, 0x10 },
+	{ 0x3230, CRL_REG_LEN_16BIT, 0x0312, 0x10 },
+	{ 0x3232, CRL_REG_LEN_16BIT, 0x0532, 0x10 },
+	{ 0x3234, CRL_REG_LEN_16BIT, 0x0752, 0x10 },
+	{ 0x3236, CRL_REG_LEN_16BIT, 0x00F2, 0x10 },
+	{ 0x3566, CRL_REG_LEN_16BIT, 0x3328, 0x10 },
+	{ 0x32D0, CRL_REG_LEN_16BIT, 0x3A02, 0x10 },
+	{ 0x32D2, CRL_REG_LEN_16BIT, 0x3508, 0x10 },
+	{ 0x32D4, CRL_REG_LEN_16BIT, 0x3702, 0x10 },
+	{ 0x32D6, CRL_REG_LEN_16BIT, 0x3C04, 0x10 },
+	{ 0x32DC, CRL_REG_LEN_16BIT, 0x370A, 0x10 },
+	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
+	{ 0x302A, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x302C, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x302E, CRL_REG_LEN_16BIT, 0x0003, 0x10 },
+	{ 0x3030, CRL_REG_LEN_16BIT, 0x004E, 0x10 },
+	{ 0x3036, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x3038, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
+	{ 0x30A2, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x30A6, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x3040, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3040, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x0008, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F2, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F2, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F2, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3044, CRL_REG_LEN_16BIT, 0x0400, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x3064, CRL_REG_LEN_16BIT, 0x1802, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x3180, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
+	{ 0x33E4, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x33E0, CRL_REG_LEN_16BIT, 0x0C80, 0x10 },
+	{ 0x3004, CRL_REG_LEN_16BIT, 0x0004, 0x10 },
+	{ 0x3008, CRL_REG_LEN_16BIT, 0x0783, 0x10 },
+	{ 0x3002, CRL_REG_LEN_16BIT, 0x0040, 0x10 },
+	{ 0x3006, CRL_REG_LEN_16BIT, 0x0477, 0x10 },
+	{ 0x3032, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x3400, CRL_REG_LEN_16BIT, 0x0010, 0x10 },
+	{ 0x3402, CRL_REG_LEN_16BIT, 0x0788, 0x10 },
+	{ 0x3402, CRL_REG_LEN_16BIT, 0x0F10, 0x10 },
+	{ 0x3404, CRL_REG_LEN_16BIT, 0x04B8, 0x10 },
+	{ 0x3404, CRL_REG_LEN_16BIT, 0x0970, 0x10 },
+	{ 0x3082, CRL_REG_LEN_16BIT, 0x000C, 0x10 },
+	{ 0x30BA, CRL_REG_LEN_16BIT, 0x11F3, 0x10 },
+	{ 0x300C, CRL_REG_LEN_16BIT, 0x08C6, 0x10 },
+	{ 0x300A, CRL_REG_LEN_16BIT, 0x0518, 0x10 },
 	{ 0x3042, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
 	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
-	{ 0x3012, CRL_REG_LEN_16BIT, 0x0163, 0x10 },
-	{ 0x3014, CRL_REG_LEN_16BIT, 0x014F, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3238, CRL_REG_LEN_16BIT, 0x0222, 0x10 },
+	{ 0x3012, CRL_REG_LEN_16BIT, 0x0131, 0x10 },
+	{ 0x3014, CRL_REG_LEN_16BIT, 0x098E, 0x10 },
+	{ 0x321E, CRL_REG_LEN_16BIT, 0x098E, 0x10 },
+	{ 0x3222, CRL_REG_LEN_16BIT, 0x098E, 0x10 },
+	{ 0x3226, CRL_REG_LEN_16BIT, 0x098E, 0x10 },
 	{ 0x30B0, CRL_REG_LEN_16BIT, 0x0800, 0x10 },
-	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C08, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C0E, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C0E, 0x10 },
+	{ 0x32EA, CRL_REG_LEN_16BIT, 0x3C0E, 0x10 },
 	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
-	{ 0x31D0, CRL_REG_LEN_16BIT, 0x0000, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x32EC, CRL_REG_LEN_16BIT, 0x72A1, 0x10 },
+	{ 0x31D0, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
+	{ 0x31AE, CRL_REG_LEN_16BIT, 0x0201, 0x10 },
 	{ 0x31AE, CRL_REG_LEN_16BIT, 0x0001, 0x10 },
-	{ 0x31AC, CRL_REG_LEN_16BIT, 0x0C0C, 0x10 },
-	/* try sync mode */
+	{ 0x31AC, CRL_REG_LEN_16BIT, 0x140C, 0x10 },
 	{ 0x340A, CRL_REG_LEN_16BIT, 0x0077, 0x10 },
 	{ 0x340C, CRL_REG_LEN_16BIT, 0x0080, 0x10 },
 	{ 0x30CE, CRL_REG_LEN_16BIT, 0x0120, 0x10 },
@@ -394,10 +1433,55 @@ struct crl_mode_rep ar0231at_modes[] = {
 		.height = 1080,
 		.min_llp = 2162,
 		.min_fll = 1354,
-		.comp_items = 0,
-		.ctrl_data = 0,
-		.mode_regs_items = ARRAY_SIZE(ar0231at_1920_1080),
-		.mode_regs = ar0231at_1920_1080,
+		.comp_items = 1,
+		.ctrl_data = &ar0231at_ctrl_data_modes[0],
+		.mode_regs_items = ARRAY_SIZE(ar0231at_1920_1080_linear_mode),
+		.mode_regs = ar0231at_1920_1080_linear_mode,
+	},
+	{
+		.sd_rects_items = ARRAY_SIZE(ar0231at_1920_1080_rects),
+		.sd_rects = ar0231at_1920_1080_rects,
+		.binn_hor = 1,
+		.binn_vert = 1,
+		.scale_m = 1,
+		.width = 1920,
+		.height = 1080,
+		.min_llp = 1978,
+		.min_fll = 1480,
+		.comp_items = 1,
+		.ctrl_data = &ar0231at_ctrl_data_modes[1],
+		.mode_regs_items = ARRAY_SIZE(ar0231at_1920_1080_2hdr_mode),
+		.mode_regs = ar0231at_1920_1080_2hdr_mode,
+	},
+	{
+		.sd_rects_items = ARRAY_SIZE(ar0231at_1920_1080_rects),
+		.sd_rects = ar0231at_1920_1080_rects,
+		.binn_hor = 1,
+		.binn_vert = 1,
+		.scale_m = 1,
+		.width = 1920,
+		.height = 1080,
+		.min_llp = 1978,
+		.min_fll = 1480,
+		.comp_items = 1,
+		.ctrl_data = &ar0231at_ctrl_data_modes[2],
+		.mode_regs_items = ARRAY_SIZE(ar0231at_1920_1080_3hdr_mode),
+		.mode_regs = ar0231at_1920_1080_3hdr_mode,
+	},
+	{
+		.sd_rects_items = ARRAY_SIZE(ar0231at_1920_1080_rects),
+		.sd_rects = ar0231at_1920_1080_rects,
+		.binn_hor = 1,
+		.binn_vert = 1,
+		.scale_m = 1,
+		.width = 1920,
+		.height = 1080,
+		.min_llp = 2246,
+		.min_fll = 1304,
+		.comp_items = 1,
+		.ctrl_data = &ar0231at_ctrl_data_modes[3],
+		.mode_regs_items = ARRAY_SIZE(ar0231at_1920_1080_4hdr_mode),
+		.mode_regs = ar0231at_1920_1080_4hdr_mode,
 	},
 };
 
@@ -627,7 +1711,7 @@ struct crl_v4l2_ctrl ar0231at_v4l2_ctrls[] = {
 		.data.std_data.min = 1920,
 		.data.std_data.max = 65535,
 		.data.std_data.step = 1,
-		.data.std_data.def = 2162,
+		.data.std_data.def = 1978,
 		.flags = V4L2_CTRL_FLAG_UPDATE,
 		.impact = CRL_IMPACTS_NO_IMPACT,
 		.ctrl = 0,
@@ -647,7 +1731,7 @@ struct crl_v4l2_ctrl ar0231at_v4l2_ctrls[] = {
 		.data.std_data.min = 1080,
 		.data.std_data.max = 65535,
 		.data.std_data.step = 1,
-		.data.std_data.def = 1354,
+		.data.std_data.def = 1480,
 		.flags = V4L2_CTRL_FLAG_UPDATE,
 		.impact = CRL_IMPACTS_NO_IMPACT,
 		.ctrl = 0,
@@ -707,7 +1791,7 @@ struct crl_v4l2_ctrl ar0231at_v4l2_ctrls[] = {
 		.data.std_data.step = 1,
 		.data.std_data.def = 0x0,
 		.flags = V4L2_CTRL_FLAG_UPDATE,
-		.impact = CRL_IMPACTS_NO_IMPACT,
+		.impact = CRL_IMPACTS_MODE_SELECTION,
 		.ctrl = 0,
 		.regs_items = ARRAY_SIZE(ar0231at_exposure_mode_regs),
 		.regs = ar0231at_exposure_mode_regs,
-- 
1.9.1

